domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandlerDebugImpl.kt

```diff
@@ -0,0 +1,126 @@
+package org.oppia.android.domain.hintsandsolution
+
+import org.oppia.android.app.model.HelpIndex
+import org.oppia.android.app.model.State
+import org.oppia.android.domain.devoptions.ShowAllHintsAndSolutionHelper
+import java.util.concurrent.locks.ReentrantLock
+import javax.inject.Inject
+import kotlin.concurrent.withLock
+
+/**
+ * Debug implementation of [HintHandler] that overrides the default mechanism of hints & solutions
+ * to show all hints and solutions based on availability, bypassing the wrong answer check and
+ * scheduling of hints, if 'Show all hints and solution' functionality is enabled in the
+ * 'Developer Options Menu'. If this functionality is disabled then it will fall back to
+ * [HintHandlerProdImpl] and provide its functionalities.
+ */
+class HintHandlerDebugImpl private constructor(
+  private val hintHandlerProdImpl: HintHandlerProdImpl,
+  private val showAllHintsAndSolutionHelper: ShowAllHintsAndSolutionHelper,
+  private val hintMonitor: HintHandler.HintMonitor
+) : HintHandler {
+
+  private val handlerLock = ReentrantLock()
+
+  override fun startWatchingForHintsInNewState(state: State) {
+    if (!showAllHintsAndSolutionHelper.getShowAllHintsAndSolution()) {
```

Comment:
Should we maybe cache this value upon debug impl creation? Otherwise, we might have some whacky behavior if it's ever possible to change this state mid-exploration.

It also could largely simplify this implementation: consider querying for this state in the factory and providing either the debug or prod implementation (which would let you remove all the prod impl calls from the debug version & just perform the debug-only functionality).

Reply:
We can't change the state mid-exploration as the developer options menu is not accessible while in an exploration. However, I think we can go with the factory approach you suggested. It will simplify the implementation even further.

Reply:
Done. PTAL. Thanks.